ext/ffi/callback.rs

```diff
@@ -121,51 +141,50 @@ unsafe extern "C" fn deno_ffi_callback(
 ) {
   LOCAL_THREAD_ID.with(|s| {
     if *s.borrow() == info.thread_id {
-      // Own isolate thread, okay to call directly
-      do_ffi_callback(cif, info, result, args);
+      // Call from main thread. If this callback is being triggered due to a
+      // function call coming from Deno itself, then this callback will build
+      // ontop of that stack.
+      // If this callback is being triggered outside of Deno (for example from a
+      // signal handler) then this will either create an empty new stack if
+      // Deno currently has nothing running and is waiting for promises to resolve,
+      // or will (very incorrectly) build ontop of whatever stack exists.
+      // The callback will even be called through from a `while (true)` liveloop, but
+      // it somehow cannot change the values that the loop sees, even if they both
+      // refer the same `let bool_value`.
+      let context: NonNull<v8::Context> = info.context;
+      let context = std::mem::transmute::<
+        NonNull<v8::Context>,
+        v8::Local<v8::Context>,
+      >(context);
+      let mut cb_scope = v8::CallbackScope::new(context);
+      let scope = &mut v8::HandleScope::new(&mut cb_scope);
+
+      do_ffi_callback(scope, cif, info, result, args);
     } else {
       let async_work_sender = &info.async_work_sender;
-      // SAFETY: Safe as this function blocks until `do_ffi_callback` completes and a response message is received.
-      let cif: &'static libffi::low::ffi_cif = std::mem::transmute(cif);
-      let result: &'static mut c_void = std::mem::transmute(result);
-      let info: &'static CallbackInfo = std::mem::transmute(info);
-      let (response_sender, response_receiver) = sync_channel::<()>(0);
-      let fut = Box::new(move || {
-        do_ffi_callback(cif, info, result, args);
-        response_sender.send(()).unwrap();
+
+      let mut args = TaskArgs {
+        cif: NonNull::from(cif),
+        result: NonNull::from(result),
+        args,
+        info: NonNull::from(info),
+      };
+
+      async_work_sender.spawn_blocking(move |scope| {
+        args.run(scope);
```

Comment:
**question:** What happens with errors thrown from here?

Reply:
Because these are tasks and we're technically violating the contract by allowing them to bubble into the scope, they end up as "uncaught errors". It doesn't kill the runtime but it isn't a great situation.

It should not be the same as our current state of affairs, however, but I suppose we could add a `tc_scope` here and print a last-ditch message.

Reply:
Note that we will need to figure out what to do with this test as it's failing right now -- either we could punt with a TODO for later, or try to add the last-ditch handler.

Reply:
Hm. I did add the error and unhandled rejection listeners in the test but they were not getting fired. Are you sure the errors end up as properly handled uncaught errors/rejections?

Reply:
Let me dig into this a bit and see what is going on before we land it...